src/lib.rs

```diff
@@ -84,3 +126,97 @@ fn load_pem_certs(path: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
         })
         .collect()
 }
+
+/// Load certificate from certificate directory (what OpenSSL calls CAdir)
+///
+/// This directory can contain other files and directories. CAfile tends
+/// to be in here too. To avoid loading something twice or something that
+/// isn't an valid certificate. Load is limited to those files that have
+/// a hash-based file name matching the pattern used by OpenSSL. The hash
+/// is not verified, however.
+fn load_pem_certs_from_dir(dir: &Path) -> Result<Vec<CertificateDer<'static>>, Error> {
+    let dir_reader = fs::read_dir(dir)?;
+    let mut certs = Vec::new();
+    for entry in dir_reader {
+        let entry = entry?;
+        let path = entry.path();
+        let file_name = path
+            .file_name()
+            .expect("dir entry with no name");
+        let metadata = match fs::metadata(&path) {
+            Ok(metadata) => metadata,
+            Err(e) if e.kind() == ErrorKind::NotFound => {
+                // Dangling symlink
+                continue;
+            }
+            Err(e) => return Err(e),
+        };
+        if metadata.is_file() && is_hash_file_name(file_name) {
+            certs.append(&mut load_pem_certs(&path)?);
+        }
+    }
+    Ok(certs)
+}
+
+/// Check if this is a hash-based file name for a certificate
+///
+/// According to the [c_rehash man page][]:
+///
+/// > The links created are of the form HHHHHHHH.D, where each H is a hexadecimal
+/// > character and D is a single decimal digit.
+///
+/// `c_rehash` generates lower-case hex digits but this is not clearly documented.
+/// Because of this, and because it could lead to issues on case-insensitive file
+/// systems, upper-case hex digits are accepted too.
+///
+/// [c_rehash man page]: https://www.openssl.org/docs/manmaster/man1/c_rehash.html
+fn is_hash_file_name(file_name: &OsStr) -> bool {
+    let file_name = match file_name.to_str() {
+        Some(file_name) => file_name,
+        None => return false, // non-UTF8 can't be hex digits
+    };
+
+    if file_name.len() == 10 {
+        let mut iter = file_name.chars();
+        let iter = iter.by_ref();
+        iter.take(8)
+            .all(|c| c.is_ascii_hexdigit())
+            && iter.next() == Some('.')
+            && matches!(iter.next(), Some('0'..='9'))
```

Comment:
optional nit: I think this last `matches!` might be a touch more readable using `is_ascii_digit` like:
```
&& matches!(iter.next(), Some(c) if c.is_ascii_digit())
```

Reply:
Done